src/computedevs.jl

```diff
@@ -0,0 +1,160 @@
+
+adapt_storage(::CPUDevice, x) = adapt_storage(Array, x)
+
+Sys.total_memory(::CPUDevice) = Sys.total_memory()
+Sys.free_memory(::CPUDevice) = Sys.free_memory()
+
+
+"""
+    abstract type AbstractComputeAccelerator <: AbstractComputeUnit
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractComputeAccelerator <: AbstractComputeUnit end
+export AbstractComputeAccelerator
+
+
+"""
+    abstract type AbstractGPUDevice <: AbstractComputeAccelerator
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputeAccelerator end
+export AbstractGPUDevice
+
+
+merge_compute_units() = ComputingDeviceIndependent()
+
+@inline function merge_compute_units(a, b, c, ds::Vararg{Any,N}) where N
+    a_b = merge_compute_units(a,b)
+    return merge_compute_units(a_b, c, ds...)
+end
+
+@inline merge_compute_units(a::UnknownComputeUnitOf, b::UnknownComputeUnitOf) = a
+@inline merge_compute_units(a::UnknownComputeUnitOf, b::Any) = a
+@inline merge_compute_units(a::Any, b::UnknownComputeUnitOf) = b
+
+@inline function merge_compute_units(a, b)
+    return (a === b) ? a : compute_unit_mergeresult(
+        compute_unit_mergerule(a, b),
+        compute_unit_mergerule(b, a),
+    )
+end
+
+struct NoCUnitMergeRule end
+
+@inline compute_unit_mergerule(a::Any, b::Any) = NoCUnitMergeRule()
+@inline compute_unit_mergerule(a::UnknownComputeUnitOf, b::Any) = a
+@inline compute_unit_mergerule(a::UnknownComputeUnitOf, b::UnknownComputeUnitOf) = a
+@inline compute_unit_mergerule(a::ComputingDeviceIndependent, b::Any) = b
+
+@inline compute_unit_mergeresult(a_b::NoCUnitMergeRule, b_a::NoCUnitMergeRule) = MixedComputeSystem()
+@inline compute_unit_mergeresult(a_b, b_a::NoCUnitMergeRule) = a_b
+@inline compute_unit_mergeresult(a_b::NoCUnitMergeRule, b_a) = b_a
+@inline compute_unit_mergeresult(a_b, b_a) = a_b === b_a ? a_b : MixedComputeSystem()
+
+
+"""
+    get_compute_unit(x)::Union{
+        AbstractComputeUnit,
+        ComputingDeviceIndependent,
+        UnknownComputeUnitOf
+    }
+
+Get the computing device backing object `x`.
+
+Don't specialize `get_compute_unit`, specialize
+[`Adapt.get_compute_unit_impl`](@ref) instead.
+"""
+get_compute_unit(x) = get_compute_unit_impl(Union{}, x)
+export get_compute_unit
+
+
+"""
+    get_compute_unit_impl(::Type{TypeHistory}, x)::AbstractComputeUnit
+
+See [`get_compute_unit_impl`](@ref).
+
+Specializations that directly resolve the compute unit based on `x` can
+ignore `TypeHistory`:
+
+```julia
+Adapt.get_compute_unit_impl(@nospecialize(TypeHistory::Type), x::SomeType) = ...
+```
+"""
+function get_compute_unit_impl end
+
+
+@inline get_compute_unit_impl(@nospecialize(TypeHistory::Type), ::Array) = CPUDevice()
+
+# Guard against object reference loops:
+@inline get_compute_unit_impl(::Type{TypeHistory}, x::T) where {TypeHistory,T<:TypeHistory} = begin
+    UnknownComputeUnitOf(x) 
+end
+
+@generated function get_compute_unit_impl(::Type{TypeHistory}, x) where TypeHistory
```

Comment:
I am really not a fan of this `@generated` implementation. I think it would be preferable to follow the Adapt.jl pattern here and to perform a tree-walk. Now it is probably the case that the current infrastructure is not general enough, but if this goes into Adapt.jl we should have one mechanism to do this.

@maleadt and I briefly spoke about this and in general we are not opposed to this functionality being in Adapt.jl

Reply:
> @maleadt and I briefly spoke about this and in general we are not opposed to this functionality being in Adapt.jl

Thanks!

Reply:
> I am really not a fan of this @generated implementation [...] would be preferable to follow the Adapt.jl pattern here and to perform a tree-walk [...] if this goes into Adapt.jl we should have one mechanism to do this.

Do you mean that it's `@generated`, or what it does? In principle it does a tree walk, I would say (and object's encountered on the way can use a different method/specialization of `get_compute_unit_impl`. The function could be written without `@generated`, it could be done with just `getfield`, `map` and `ntuple`. I only wrote it this way to minimize the resulting code and increase type stability. But maybe I'm missing your point here?

Reply:
There are two points:

1. I personally aim to reduce reliance on `@generated` as much as possible. So if we can write this without staged functions that would be great, it often makes the intent clearer and simplifies extendability (and is better for compile times, also better type-stability)
2. This duplicates the core functionality of Adapt.jl which is essentially a tree walk over structs, so instead of two implementations it would be great to have one, but Tim and I acknowledge that this might be a trickier design.

Reply:
Ok, I'll try to rewrite without `@generated`.

Reply:
> without staged functions [...] better type-stability

Type stability was actually one of the reasons I went for `@generated` here. :-)

Ok I tried without `@generated` @vchuravy, but type stability suffers. One implementation I came up with is

```
function get_compute_unit_impl_nogen(::Type{TypeHistory}, x::T) where {TypeHistory,T}
    NewTypeHistory = Union{TypeHistory, T}
    fields = ntuple(Base.Fix1(getfield, x), Val(fieldcount(T)))
    merge_compute_units(map(Base.Fix1(get_compute_unit_impl, NewTypeHistory), fields)...)
end
```

Nice and short, but:

```julia
julia> x = (a = 4.2, b = (c = rand(2,3)))
(a = 4.2, b = [0.7927795137326867 0.8930673224466184 0.15921059563423712; 0.6399176439174568 0.4501022168243579 0.3951239506670382])

julia> @inferred get_compute_unit_impl(Union{}, x)
Adapt.CPUDevice()

julia> @inferred get_compute_unit_impl_nogen(Union{}, x)
ERROR: return type Adapt.CPUDevice does not match inferred return type Any
```

And in truth, this version also uses generated code underneath, because `ntuple` does. Is there a way to do this without `ntuple`, efficiently?